url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/ExecStageThumb.java

```diff
@@ -0,0 +1,2318 @@
+        int rn = inst.getField(3, 3);
+        int rd = inst.getRdField();
+        int vaddr, paddr;
+
+        if (!exec) {
+            printDisasm(inst, "strh",
+                    String.format("%s, [%s, %s]",
+                            getRegName(rd), getRegName(rn), getRegName(rm)));
+            return;
+        }
+
+        vaddr = getReg(rn) + getReg(rm);
+
+        paddr = getMMU().translate(vaddr, 2, false, getCPSR().isPrivMode(), false);
+        if (getMMU().isFault()) {
+            getMMU().clearFault();
+            return;
+        }
+
+        if (!tryWrite_a32(paddr, 2)) {
+            raiseException(ARMv5.EXCEPT_ABT_DATA,
+                    String.format("strh [%08x]", paddr));
+            return;
+        }
+        write16_a32(paddr, (short) getReg(rd));
+    }
+
+    /**
+     * プッシュ命令。
+     *
+     * @param inst Thumb 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executePush(InstructionThumb inst, boolean exec) {
+        boolean br = inst.getBit(8);
+        int rlist = inst.getRegListField();
+        int vaddr, paddr, len;
+
+        if (!exec) {
+            printDisasm(inst, "push",
+                    String.format("{%s%s%s}",
+                            inst.getRegListFieldName(),
+                            (inst.getRegListField() != 0 && br) ? ", " : "",
+                            (br) ? "lr" : ""));
+            return;
+        }
+
+        vaddr = getReg(13) - (Integer.bitCount(rlist) * 4);
+        len = -(Integer.bitCount(rlist) * 4);
+        if (br) {
+            vaddr -= 4;
+            len -= 4;
+        }
+        for (int i = 0; i < 8; i++) {
+            if ((rlist & (1 << i)) == 0) {
+                continue;
+            }
+
+            paddr = getMMU().translate(vaddr, 4, false, getCPSR().isPrivMode(), false);
+            if (getMMU().isFault()) {
+                getMMU().clearFault();
+                return;
+            }
+
+            if (!tryWrite_a32(paddr, 4)) {
+                raiseException(ARMv5.EXCEPT_ABT_DATA,
+                        String.format("push [%08x]", paddr));
+                return;
+            }
+            write32_a32(paddr, getReg(i));
+            vaddr += 4;
+        }
+        if (br) {
+            paddr = getMMU().translate(vaddr, 4, false, getCPSR().isPrivMode(), false);
+            if (getMMU().isFault()) {
+                getMMU().clearFault();
+                return;
+            }
+
+            if (!tryWrite_a32(paddr, 4)) {
+                raiseException(ARMv5.EXCEPT_ABT_DATA,
+                        String.format("push [%08x]", paddr));
+                return;
+            }
+            write32_a32(paddr, getReg(14));
+            vaddr += 4;
+        }
+
+        setReg(13, getReg(13) + len);
+    }
+
+    /**
+     * ポップ命令。
+     *
+     * @param inst Thumb 命令
+     * @param exec デコードと実行なら true、デコードのみなら false
+     */
+    public void executePop(InstructionThumb inst, boolean exec) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code block on lines 1879-1898 into a separate method.